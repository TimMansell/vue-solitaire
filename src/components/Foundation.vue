<template>
  <div class="foundation">
    <div
      class="foundation__column"
      v-for="(foundation, foundationsIndex) in foundationCards"
      :key="`f-${foundationsIndex}`"
      @click="setFoundationColumn(foundationsIndex)"
      :data-test="`foundation-${foundationsIndex}`"
      @drop="dropCard(foundationsIndex)"
      @dragover.prevent
      @dragenter.prevent
    >
      <Card
        v-for="(card, foundationIndex) in foundation"
        :key="`a-${foundationIndex}`"
        :value="card.value"
        :suit="card.suit"
        :visible="card.visible"
        :clickable="false"
        stacked
      />

      <CardPlaceholder v-if="!foundation.length" see-through />
    </div>
  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex';
import Card from '@/components/Card.vue';
import CardPlaceholder from '@/components/CardPlaceholder.vue';

export default {
  name: 'Foundation',
  components: {
    Card,
    CardPlaceholder,
  },
  computed: {
    ...mapGetters(['foundationCards']),
  },
  methods: {
    ...mapActions(['moveCardToFoundation']),
    setFoundationColumn(columnIndex) {
      this.moveCardToFoundation(columnIndex);
    },
    dropCard(columnIndex) {
      this.moveCardToFoundation(columnIndex);
    },
  },
};
</script>

<style scoped lang="scss">
.foundation {
  display: flex;
  margin-bottom: var(--mg-sm);
  margin-left: calc(var(--col-padding) * -1);
  margin-right: calc(var(--col-padding) * -1);

  @media (min-width: $bp-md) {
    margin-bottom: var(--mg-md);
    margin-left: calc(var(--col-padding-lg) * -1);
    margin-right: calc(var(--col-padding-lg) * -1);
  }

  &__column {
    padding-left: var(--col-padding);
    padding-right: var(--col-padding);

    @media (min-width: $bp-md) {
      padding-left: var(--col-padding-lg);
      padding-right: var(--col-padding-lg);
    }
  }
}
</style>
