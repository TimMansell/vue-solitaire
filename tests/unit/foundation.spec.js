import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Foundation from '@/components/Foundation.vue';
// import state from '@/store/state';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('Foundation.vue', () => {
  it('matches snapshot', () => {
    const getters = {
      foundationCards: () => ([{
        value: 'A',
        order: 1,
        suit: 's',
        position: [7, 5],
        visible: true,
      }]),
    };

    const store = new Vuex.Store({
      getters,
    });

    const wrapper = shallowMount(Foundation, { store, localVue });

    expect(wrapper).toMatchSnapshot();
  });

  it('calls store action "moveCardToFoundation" when clicked', () => {
    const getters = {
      foundationCards: () => ([{
        value: 'A',
        order: 1,
        suit: 's',
        position: [7, 5],
        visible: true,
      }]),
    };

    const actions = {
      moveCardToFoundation: jest.fn(),
    };

    const store = new Vuex.Store({
      getters,
      actions,
    });

    const wrapper = shallowMount(Foundation, { store, localVue });

    wrapper.find('[data-test="foundation-0"]').trigger('click');

    expect(actions.moveCardToFoundation).toHaveBeenCalled();
  });
});
