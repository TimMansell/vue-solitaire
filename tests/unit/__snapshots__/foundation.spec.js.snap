// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Foundation.vue matches snapshot 1`] = `
<div data-test="ace" class="foundation">
  <div data-test="foundation-0" class="foundation__column">
    <card-stub value="A" suit="c" order="0" position=""></card-stub>
    <card-stub value="A" suit="c" order="0" position=""></card-stub>
    <card-stub value="A" suit="c" order="0" position=""></card-stub>
    <card-stub value="A" suit="c" order="0" position=""></card-stub>
    <card-stub value="A" suit="c" order="0" position=""></card-stub>
    <svgicon-stub name="Card_back_17" class="foundation__placeholder"></svgicon-stub>
  </div>
</div>
`;
